### PR TITLE
setup crude hot reload with nodemon

### PR DESCRIPTION
**Pull request by ahmad2smile**
#### Feature: Hot Reload

This is a very crud implementation of that should be an actual Hot Reload functionality for this project, #1  But for now I thought it could be useful. It still takes around 1 min to reload on Mac Book Pro but you don't have to do anything manually at least.

This PR sets the functionality of rebuilding of `wasm` artifacts based on any changes detected with `nodemon` in `src` directory.

The files serving setup on the localhost been removed from `Makefile` and moved to `package.json` just for `development`. `parcel` would still be serving files on `localhost` but as the HMR feature of `parcel` wasn't working I setup a hack for hard reload on HMR.

Now, `nodemon` will build `wasm` outputs with `make start` on changes in `src` and `parcel` would reload page when `nodemon` is done.

In the start after `npm install` first build is required so all the files are in place for `parcel` as i'm running `build` and `serve` command with `concurrently`.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
-  loader = import('../glue/asm.js');
+	loader = import("../glue/asm.js")
 }
 
-loader.then(module => module.default(config));
+loader.then(module => module.default(config))
+
+// Parcel right now doesn't support --reload
+// and its HMR is not working on with asm-node, hence hard-reload
+if (module.hot) {
```
Comment (mbasso) [inline src/index.js:17]: Cool, looks good to me. But what if we put this code in the files in the `glue` folder? In this way logic is hidden to the final user. Also, `index.js` might be modified with other imports that doesn't require a complete reload of the page

**Comment by mbasso** [comment]
Thank you for submitting this PR, I really appreciate your help!

I think that `concurrently` is the right tool to do that.
However I've a proposal to replace `nodemon` with a script.
At the moment we need to recompile the app when some source code files are changed, this means that we need to watch the src folder (in particular only`.cpp` files, as `parcel` will automatically detect changes to `.js`, `.css` files and so on) and the [files defined on top of the Makefile](https://github.com/mbasso/asm-dom-boilerplate/blob/master/Makefile#L11). At the moment there's only the source code of asm-dom but that param can be changed manually by users to add their dependencies, external libraries etc. With the current implementation changes to these files will not trigger a compilation, so, a reload of the app.
To solve this issue I think that we can take advantage of the cache of `Make`. As shown [here](https://stackoverflow.com/a/48496716) we can use a simple command to run make every second, for example, and recompile the app only if some files are changed. In this way we can also work on/with external libs without restart the development server.
To do that I think that we need to [fix the command to build .bc file](https://github.com/mbasso/asm-dom-boilerplate/blob/master/Makefile#L86), we have to add `FILES` to prerequisites:

```
$(BC): $(OBJS) $(FILES)
  emcc \
    $(CFLAGS) \
    --bind \
    $(INCLUDES) \
    $(OBJS) \
    $(FILES) \
    -o $(BC)
```

and change the start command in the `MakeFile` like this:

```
# this on top of MakeFile after CPPFLAGS
START_PREREQUISITES := \
  dist/wasm \
  dist/asmjs

# this replace the current "start" command
# -p, -n and -c just add some colors and information about the commands runned by concurrently
# I've set a timer of 1.5s by default
start: $(START_PREREQUISITES)
  npx concurrently \
    -p "[{name}]" \
    -n "Parcel,Make" \
    -c "bgGreen.bold,bgBlue.bold" \
    "npx cross-env NODE_ENV=development parcel index.html" \
    "while true; do make -q $(START_PREREQUISITES) || make $(START_PREREQUISITES); sleep 1.5; done"
```

The problem here is that I don't know if this command is compatible with windows, alternatively we can find an `npm` package or change the script based on platform

```
while true; do make -q $(START_PREREQUISITES) || make $(START_PREREQUISITES); sleep 1.5; done
```

What do you think about that? Does this make sense to you?

File: src/index.js (hunk 1 above)
Comment (ahmad2smile) [reply src/index.js:17]: If you move this setup to glue dir in any files there, page doesn't reload. some exception throw regarding `tagName` of undefined besides this setup is just for the parcel to override the default HMR behavior. It has nothing to do with the scope of the files to be reloaded as that is handled by parcel.

**Comment by ahmad2smile** [comment]
I've tested your Makefile script on Mac its working fine, although pooling setup of it rather than watching for file changes kinda bugs me. Still, I'm gonna check on Windows 10 Dell PC. I'll keep you posted of any progress.

**Comment by mbasso** [comment]
yeah, that sounds strange to me too, a watcher is definitely a cleaner solution, but it seems that there is no way to watch all prerequisites of a `make` command, we need a specific tool to do that.
I've found [this](https://github.com/andrewffff/watch-make) repository, I haven't read the source code yet, but it seems that it uses an interesting mechanism to watch only the files that are needed by a `make` command. Unfortunately this project seems a little bit outdated and unmaintained...

**Comment by ahmad2smile** [comment]
Tested the make script on Mac Book Pro High Sierra. But was unable to test on Window cause of #3

**Comment by mbasso** [comment]
I'll try to solve it as soon as possible!
Meanwhile I've tried to write a cli based on my previous comment, it is not completed yet but I think that it might be useful to a lot of people, [here](https://github.com/mbasso/make-watch) is the repository if you want to contribute

**Comment by ahmad2smile** [comment]
Ok. I checked the new Make script on Windows 10. Its not completely working.

1. The error on [Cmder](http://cmder.net/)
```
ERROR: Invalid argument/option - 'P:/'.
Type "FORFILES /?" for usage.
emcc \
        -O3 -Wall -Werror -Wall -Wno-deprecated -Wno-parentheses -Wno-format \
        --bind \
        -include node_modules/asm-dom/cpp/asm-dom.hpp \
         \
        node_modules/asm-dom/cpp/asm-dom.cpp \
        -o temp/app.bc
error: unable to open output file 'temp/app.bc': 'no such file or directory'
1 error generated.
ERROR:root:compiler frontend failed to generate LLVM bitcode, halting
make: *** [makefile:92: temp/app.bc] Error 1
```

2. Error on Regular `cmd`. Though I runs the project and builds fine it doesn't watch/pool.
```
[ Make] 'while' is not recognized as an internal or external command,
[ Make] operable program or batch file.
```

**Comment by mbasso** [comment]
> The error on Cmder

Unfortunately I'm unable to reproduce the first error. Could you please try to run that command manually replacing the `$(SRCDIR)` with its value? I've just pushed a potential fix, let me know if it works

> Error on Regular cmd. Though I runs the project and builds fine it doesn't watch/pool.

I think that it is not compatible with windows, we have to write another version of the script to use on windows.
Alternatively, I think that we can use directly [make-watch](https://github.com/mbasso/make-watch),  We need only to finish tests and try it on a real project like this.

**Comment by ahmad2smile** [comment]
I'll check the new update. Regarding `make-watch` I checked you are using `chokidar` package.  I was thinking maybe we should switch this `while` setup with [watchman](https://github.com/facebook/watchman) and maybe you can check it out for `make-watch` too. Its by facebook.

**Comment by mbasso** [comment]
Wow, I didn't know it, seems interesting, I have to learn more about that. Thank you!

Yeah, I think that we can use a watcher until `make-watch` is not completed (it will allow to watch also the `Makefile` itself and other dependencies outside the src folder extracted directly from the `make` command). However it seems that `watchman` need to be installed on the OS and only after that it can be used from node, I think that this might not be the best for the users. There is also [an issue](https://github.com/parcel-bundler/parcel/issues/863) on [parcel repo](https://github.com/parcel-bundler/parcel/issues/863). I think that is better to achieve the same result only with an `npm install`, so, `nodemon` might be the right choice, as you had already done... just add the config file and replace the `while` script with something like `nodemon --exec 'make $(START_PREREQUISITES)'` should work fine without moving logic to `package.json`

**Comment by ahmad2smile** [comment]
Replaced `while` logic with `nodemon` in `Makefile`. Tested on Mac Book Pro, Windows testing is pending.

**Comment by mbasso** [comment]
I've tried it on Windows and it seems to work well. I only had to update the `Makefile`  with the latest changes on master (to fix some issues on windows) and change a little bit the command:

```bash
# from this one
"nodemon --exec 'make $(START_PREREQUISITES)'"

# to this one (with double quotes)
"nodemon --exec \"make $(START_PREREQUISITES)\""
````

I haven't tested it on Mac yet.

**Comment by ahmad2smile** [comment]
I've updated the command to use double quotes with escape characters. I'v tested on Mac and it works fine.

**Comment by mbasso** [comment]
awesome! I'll merge and close #1, thank you so much for contributing!

**Pull request merged**